### PR TITLE
🐛Source-postgres: added materialized views processing

### LINKED ISSUE #9012: Postgres Source Connector Hides Materialized Views
## Environment
- **Airbyte version**:  example is 0.32.5-alpha
- **OS Version / Instance**: Amazon Linux 2, AWS EC2
- **Deployment**: Docker
- **Source Connector and version**: 0.3.14+
- **Destination Connector and version**: N/A
- **Severity**: Very Low / Low / Medium / High / Critical
- **Step where error happened**: Setup new connection / Update latest source schema

## Current Behavior
Materialized views don't show up in the source schema due to this change: https://github.com/airbytehq/airbyte/pull/8010 v0.3.13 and earlier worked fine, but after 0.3.14 materialized views don't show up. Views and tables work fine, but materialized views don't. Here's a StackOverflow post I found that describes how getting privileges for materialized views is different from regular views and tables: https://stackoverflow.com/questions/38728548/list-grants-and-privileges-for-a-materialized-view-in-postgresql

## Expected Behavior
Materialized views should be included in the source schema catalog

## Logs
N/A

## Steps to Reproduce
1. In the source database, create a materialized view (e.g. `CREATE MATERIALIZED VIEW my_table AS ...`)
2. In Airbyte click "Update source schema" button and the materialized view won't be in the list

## Are you willing to submit a PR?

Remove this with your answer.

### PR DESCRIPTION
**Pull request by etsybaev**
## What
Currently, we are not able to read Materialized views.

## How

Updated connector to make it able to process Materialized views.

Tested locally.
Created some test Materialized views in dockerized Postgres DB and synced it to local CSV destination.
Selection_316
Selection_318
Selection_317

Also checked by integrationTests:

Added new materialized view along with other tables
Selection_323
Checked that results are actually returned as expected in debug mode
Selection_321

## 🚨 User Impact 🚨
No breaking changes are expected.

## Pre-merge Checklist
Expand the relevant checklist and delete the others. 

<details><summary> <strong> New Connector </strong></summary>
<p>

#### Community member or Airbyter
   
- [ ] **Community member?** Grant edit access to maintainers ([instructions](https://docs.github.com/en/github/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork#enabling-repository-maintainer-permissions-on-existing-pull-requests))
- [ ] Secrets in the connector's spec are annotated with `airbyte_secret` 
- [ ] Unit & integration tests added and passing. Community members, please provide proof of success locally e.g: screenshot or copy-paste unit, integration, and acceptance test output. To run acceptance tests for a Python connector, follow instructions in the README. For java connectors run `./gradlew :airbyte-integrations:connectors:<name>:integrationTest`.
- [ ] Code reviews completed
- [ ] Documentation updated 
    - [ ] Connector's `README.md`
    - [ ] Connector's `bootstrap.md`. See [description and examples](https://docs.google.com/document/d/1ypdgmwmEHWv-TrO4_YOQ7pAJGVrMp5BOkEVh831N260/edit?usp=sharing)
    - [ ] `docs/SUMMARY.md`
    - [ ] `docs/integrations/<source or destination>/<name>.md` including changelog. See changelog [example](https://docs.airbyte.io/integrations/sources/stripe#changelog)
    - [ ] `docs/integrations/README.md`
    - [ ] `airbyte-integrations/builds.md`
- [ ] PR name follows [PR naming conventions](https://docs.airbyte.io/contributing-to-airbyte/updating-documentation#issues-and-pull-requests)
   
#### Airbyter

If this is a community PR, the Airbyte engineer reviewing this PR is responsible for the below items. 
   
- [ ] Create a non-forked branch based on this PR and test the below items on it
- [ ] Build is successful
- [ ] Credentials added to Github CI. [Instructions](https://docs.airbyte.io/connector-development#using-credentials-in-ci). 
- [ ] [`/test connector=connectors/<name>` command](https://docs.airbyte.io/connector-development#updating-an-existing-connector) is passing. 
- [ ] New Connector version released on Dockerhub by running the `/publish` command described [here](https://docs.airbyte.io/connector-development#updating-an-existing-connector)
- [ ] After the connector is published, connector added to connector index as described [here](https://docs.airbyte.io/connector-development#publishing-a-connector)
- [ ] Seed specs have been re-generated by building the platform and committing the changes to the seed spec files, as described [here](https://docs.airbyte.io/connector-development#publishing-a-connector)
   
</p>
</details>

<details><summary> <strong> Updating a connector </strong></summary>
<p>
   
#### Community member or Airbyter
   
- [ ] Grant edit access to maintainers ([instructions](https://docs.github.com/en/github/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork#enabling-repository-maintainer-permissions-on-existing-pull-requests))
- [ ] Secrets in the connector's spec are annotated with `airbyte_secret` 
- [x] Unit & integration tests added and passing. Community members, please provide proof of success locally e.g: screenshot or copy-paste unit, integration, and acceptance test output. To run acceptance tests for a Python connector, follow instructions in the README. For java connectors run `./gradlew :airbyte-integrations:connectors:<name>:integrationTest`.
- [x] Code reviews completed
- [x] Documentation updated 
    - [ ] Connector's `README.md`
    - [ ] Connector's `bootstrap.md`. See [description and examples](https://docs.google.com/document/d/1ypdgmwmEHWv-TrO4_YOQ7pAJGVrMp5BOkEVh831N260/edit?usp=sharing)
    - [ ] Changelog updated in `docs/integrations/<source or destination>/<name>.md` including changelog. See changelog [example](https://docs.airbyte.io/integrations/sources/stripe#changelog)
- [x] PR name follows [PR naming conventions](https://docs.airbyte.io/contributing-to-airbyte/updating-documentation#issues-and-pull-requests)
   
#### Airbyter

If this is a community PR, the Airbyte engineer reviewing this PR is responsible for the below items. 
   
- [ ] Create a non-forked branch based on this PR and test the below items on it
- [x] Build is successful
- [ ] Credentials added to Github CI. [Instructions](https://docs.airbyte.io/connector-development#using-credentials-in-ci). 
- [x] [`/test connector=connectors/<name>` command](https://docs.airbyte.io/connector-development#updating-an-existing-connector) is passing. 
- [x] New Connector version released on Dockerhub by running the `/publish` command described [here](https://docs.airbyte.io/connector-development#updating-an-existing-connector)
- [x] After the new connector version is published, connector version bumped in the seed directory as described [here](https://docs.airbyte.io/connector-development#publishing-a-connector)
- [ ] Seed specs have been re-generated by building the platform and committing the changes to the seed spec files, as described [here](https://docs.airbyte.io/connector-development#publishing-a-connector)

</p>
</details>

### CODE REVIEW

File: airbyte-integrations/connectors/source-postgres/src/main/java/io/airbyte/integrations/source/postgres/PostgresSource.java [hunk 1]
```diff
@@ -231,10 +231,29 @@ private static AirbyteStream removeIncrementalWithoutPk(final AirbyteStream stre
...
+              + "\tn.nspname as table_schema,\n"
+              + "\tc.relname as table_name,\n"
+              + "    coalesce ('SELECT') as privilege_type\n"
+              + "from \n"
+              + "    pg_class c\n"
+              + "    join pg_namespace n on n.oid = relnamespace\n"
+              + "    join pg_roles r on r.oid = relowner,\n"
+              + "    unnest(coalesce(relacl::text[], format('{%s=arwdDxt/%s}', rolname, rolname)::text[])) acl, \n"
+              + "    regexp_split_to_array(acl, '=|/') s\n"
+              + "   where r.rolname = ? and  nspname = 'public' and c.relkind = 'm' and (c.relacl is null or s[2] = 'r');");
```
Comment (alexandr-shegeda) [inline airbyte-integrations/connectors/source-postgres/src/main/java/io/airbyte/integrations/source/postgres/PostgresSource.java:253]: can we cover this change with integration test?

Comment (etsybaev) [reply airbyte-integrations/connectors/source-postgres/src/main/java/io/airbyte/integrations/source/postgres/PostgresSource.java:253]: Sure, will add it a bit later, working on tests right now

Comment (alexandertsukanov) [inline airbyte-integrations/connectors/source-postgres/src/main/java/io/airbyte/integrations/source/postgres/PostgresSource.java:253]: [OPTIONAL] As we migrate to Java 17, could we use Text Blocks https://blogs.oracle.com/javamagazine/post/text-blocks-come-to-java here? I believe it will help to make code changes faster in future.

**Review by alexandertsukanov — APPROVED** [review]
LGTM

File: airbyte-integrations/connectors/source-postgres/src/main/java/io/airbyte/integrations/source/postgres/PostgresSource.java (hunk 1 above)
Comment (etsybaev) [reply airbyte-integrations/connectors/source-postgres/src/main/java/io/airbyte/integrations/source/postgres/PostgresSource.java]: done, thanks

**Review by andriikorotkov — APPROVED** [review]

File: airbyte-integrations/connectors/source-postgres/src/main/java/io/airbyte/integrations/source/postgres/PostgresSource.java [hunk 2]
```diff
@@ -56,7 +56,8 @@ public static Source sshWrappedSource() {
   }
 
   PostgresSource() {
-    super(DRIVER_CLASS, new PostgresJdbcStreamingQueryConfiguration(), new PostgresSourceOperations());
+    super(DRIVER_CLASS, new PostgresJdbcStreamingQueryConfiguration(),
+        new PostgresSourceOperations());
```
Comment (tuliren) [inline airbyte-integrations/connectors/source-postgres/src/main/java/io/airbyte/integrations/source/postgres/PostgresSource.java:60]: Not sure if these format changes are expected. Better run `./gradle format` before merging.

File: airbyte-integrations/connectors/source-postgres/src/main/java/io/airbyte/integrations/source/postgres/PostgresSource.java [hunk 3]
```diff
@@ -228,13 +241,45 @@ private static AirbyteStream removeIncrementalWithoutPk(final AirbyteStream stre
...
     return database.query(connection -> {
       final PreparedStatement ps = connection.prepareStatement(
-          "SELECT DISTINCT table_catalog, table_schema, table_name, privilege_type\n"
-              + "FROM   information_schema.table_privileges\n"
-              + "WHERE  grantee = ? AND privilege_type = 'SELECT'");
-      ps.setString(1, database.getDatabaseConfig().get("username").asText());
+          // c.relkind = 'm' means Materialized View
+          // c.relacl is null or s[2] = 'r' means either null (all grants) or "r" - grants for read
+          // coalesce ('SELECT') as privilege_type\n" is set as it to match Union query, verification is done
+          // is "where" clause
```
Comment (tuliren) [inline airbyte-integrations/connectors/source-postgres/src/main/java/io/airbyte/integrations/source/postgres/PostgresSource.java:252]: This is really helpful.

- Can you move the comments to the SQL code directly? Like

  ```sql
  -- 'm' mean Materialized View
  AND    c.relkind = 'm'
  AND    (
                 -- all grants
                 c.relacl IS NULL
                 -- read grant
          OR     s[2] = 'r');
  ```

  In this way it is easier to tell which statement a comment is about.

- The comment about `coalesce ('SELECT') as privilege_type` is not clear. Can you rephrase it?

**Review by tuliren — APPROVED** [review]

File: airbyte-integrations/connectors/source-postgres/src/main/java/io/airbyte/integrations/source/postgres/PostgresSource.java (hunk 2 above)
Comment (etsybaev) [reply airbyte-integrations/connectors/source-postgres/src/main/java/io/airbyte/integrations/source/postgres/PostgresSource.java]: Updated, thanks!
But I always run the "format" command before pushing my changes.

File: airbyte-integrations/connectors/source-postgres/src/main/java/io/airbyte/integrations/source/postgres/PostgresSource.java (hunk 3 above)
Comment (etsybaev) [reply airbyte-integrations/connectors/source-postgres/src/main/java/io/airbyte/integrations/source/postgres/PostgresSource.java]: Updated, thanks

**Comment by etsybaev** [comment]
Since integration tests passed, and there were no new issue issues found in the build before bumping versions in dockerfile -> releasing new versions and merging

File: airbyte-integrations/connectors/source-postgres/src/main/java/io/airbyte/integrations/source/postgres/PostgresSource.java (hunk 3 above)
Comment (etsybaev) [reply airbyte-integrations/connectors/source-postgres/src/main/java/io/airbyte/integrations/source/postgres/PostgresSource.java]: /test connector=connectors/source-postgres-strict-encrypt

**Review by tuliren — APPROVED** [review]
